### PR TITLE
[Task#4937+4938] Fetch Data & Handle Event for Search Screen

### PR DESCRIPTION
**Pull request by tienpx3010**
## Fetch Data & Handle Event for Search Screen
https://edu-redmine.framgia.vn/issues/4937
https://edu-redmine.framgia.vn/issues/4938

### CODE REVIEW

File: Movie/Scenes/Home/Movie/Search/CustomSearchBar.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+//  Movie
+//
+//  Created by pham.xuan.tien on 1/9/19.
+//  Copyright © 2019 Framgia. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+protocol CustomSearchBarDelegate: NSObjectProtocol {
```
Comment (tgdong2296) [inline Movie/Scenes/Home/Movie/Search/CustomSearchBar.swift]:
```suggestion
protocol CustomSearchBarDelegate: class {
```

File: Movie/Scenes/Home/Movie/Search/SearchViewController.swift [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+                if results.isEmpty {
+                    self.movieCollectionView.showEmptyDataMessage(Constants.emptyDataMessage)
+                } else {
+                    var outputMovies = [Movie]()
+                    if self.selectedGenres.isEmpty {
+                        outputMovies = results
+                    } else {
+                        for movie in results {
+                            let resultGenres = movie.genreIds
+                            let movieSetGenres = Set(resultGenres.map({ $0 }))
```
Comment (tgdong2296) [inline Movie/Scenes/Home/Movie/Search/SearchViewController.swift]:
```suggestion
                            let movieSetGenres = Set(resultGenres.map { $0 })
```

File: Movie/Scenes/Home/Movie/Search/SearchViewController.swift [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+                            let searchGenres = Set(self.selectedGenres.map({ $0.id }))
```
Comment (tgdong2296) [inline Movie/Scenes/Home/Movie/Search/SearchViewController.swift]:
```suggestion
                            let searchGenres = Set(self.selectedGenres.map { $0.id })
```

File: Movie/Scenes/Home/Movie/Search/SearchViewController.swift [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+                            let isMatch = searchGenres.isSubset(of: movieSetGenres)
+                            if isMatch {
+                                outputMovies.append(movie)
+                            }
+                        }
+                    }
+                    if page == 1 {
+                        self.movies = outputMovies
```
Comment (tgdong2296) [inline Movie/Scenes/Home/Movie/Search/SearchViewController.swift]:
```suggestion
                        self.movies = page == 1 ? outputMovies : self.movies + outputMovies
```

File: Movie/Scenes/Home/Movie/Search/SearchViewController.swift [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        }
+    }
+
+    private func fetchGenreList() {
+        movieRepository.getGenreList { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(let genreResponse):
+                guard let results = genreResponse?.genres else { return }
+                self.genres = results.filter { $0.name != Constants.unusedCategory }
```
Comment (tgdong2296) [inline Movie/Scenes/Home/Movie/Search/SearchViewController.swift]:
```suggestion
                self.genres = results.filter { 
                             $0.name != Constants.unusedCategory 
                }
```

File: Movie/Scenes/Home/Movie/Search/CustomSearchBar.swift [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+    
+    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        guard !searchText.isEmpty else {
+            return
+        }
+        customDelegate?.customSearchBar(textWillChange: searchText)
+        throttler.throttle {
+            DispatchQueue.main.async {
```
Comment (hung070996) [inline Movie/Scenes/Home/Movie/Search/CustomSearchBar.swift]: weak self

**Pull request merged**